Source/DafnyCore/Resolver/PreType/PreTypeResolve.State.cs

```diff
@@ -0,0 +1,589 @@
+//-----------------------------------------------------------------------------
+//
+// Copyright by the contributors to the Dafny Project
+// SPDX-License-Identifier: MIT
+//
+//-----------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Diagnostics.Contracts;
+
+namespace Microsoft.Dafny {
+  public class PreTypeInferenceState {
+    public readonly PreTypeResolver PreTypeResolver;
+    private readonly DafnyOptions options;
+
+    private List<SubtypeConstraint> unnormalizedSubtypeConstraints = new();
+    private List<EqualityConstraint> equalityConstraints = new();
+    private List<Func<bool>> guardedConstraints = new();
+    private readonly List<Advice> defaultAdvice = new();
+    private List<System.Action> confirmations = new();
```

Comment:
The concept of a confirmation seems too generic, but I'm not sure how to improve it without seeing the consuming code.